### PR TITLE
refactor(api): Allow setting `ProtocolSource.directory` when reading from FS

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

This PR decouples the presence/absence of the `ProtocolSource.directory` field from how the `ProtocolSource` was computed. You can now read a protocol from files that already exist on the filesystem, *and at the same time* get a `ProtocolSource` back that has its `directory` field set. 

# Rationale

Currently, `ProtocolReader` has two modes of operation:

1. Compute a `ProtocolSource` from a bunch of file-like things, and then save those file-like things to real files on the actual filesystem.
2. Compute a `ProtocolSource` from things that are already real files on the filesystem.

In mode (1), the `directory` field is properly set on the resulting `ProtocolSource`. In mode (2), it's `None`.

The problem with this is that it's sometimes helpful for the returned `ProtocolSource` to have a `directory` *even if it was computed via mode (2).* In particular, `ProtocolStore` looks at `directory` (as well as `files`) to see what it needs to clean up when a protocol is removed. And `ProtocolStore` will be using mode (2) shortly.

In the longer term, instead of `ProtocolReader` being responsible for copying and `ProtocolStore` being responsible for deleting, we probably want a third unit that assumes both of those responsibilities. This PR doesn't do that, but I think it takes a tiny step in that direction.

# Review requests

* Please scrutinize the way I split up `ProtocolReader.read()`. I've split it into two methods that should be equivalent to the two overloads, but double-check my work.
* Can we see any worthwhile way to reduce the duplication between the two methods?

# Risk assessment

I think low, but I'm really leaning on integration tests here.

### CODE REVIEW

**Review by mcous — APPROVED** [review]
I think this looks good. As we address the remaining TODOs (and per your comment in the PR description), I think the `directory` field should move out of `ProtocolSource`

File: api/tests/opentrons/protocol_reader/test_protocol_reader.py [hunk 1]
```diff
@@ -271,10 +272,10 @@ async def test_read_files_no_copy(
     decoy.when(role_analyzer.analyze([buffered_file])).then_return(analyzed_roles)
     decoy.when(config_analyzer.analyze(main_file)).then_return(analyzed_config)
 
-    result = await subject.read(files=[input_file])
+    result = await subject.read_saved(files=[input_file], directory=directory)
 
     assert result == ProtocolSource(
-        directory=None,
+        directory=directory,
```
Comment (TamarZanzouri) [inline api/tests/opentrons/protocol_reader/test_protocol_reader.py:278]: Should we add a test for read_saved with None as a directory? I see we are testing this with legacy files but not with non legacy files. 
also, Should we test an invalid path in read_and_save? or maybe its taken care of in the file_reader_writer tests

Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_reader/test_protocol_reader.py:278]: > Should we add a test for read_saved with None as a directory?

Already did!

```python
@pytest.mark.parametrize("directory", [None, Path("/some/dir")])
```


> I see we are testing this with legacy files but not with non legacy files.
also, Should we test an invalid path in read_and_save? or maybe its taken care of in the file_reader_writer tests

Yeah, I think with the way things are currently structured:

* It's not the `ProtocolReader`'s responsibility to care about legacy vs. non-legacy files. It just passes the files straight through to its dependencies, and then coordinates between its dependencies. Possibly, we could add quasi-integration smoke tests for `ProtocolReader` the same way we have for `ProtocolRunner`.
* For the same reasons, handling the invalid path case is `FileReaderWriter`'s responsibility, as you mentioned.
